pkg/handler/metricsconfigurer/handler.go

```diff
@@ -0,0 +1,374 @@
+	// which contains the url to send the response to
+	if err := dec.Decode(&req); err != nil {
+		// make best effort to send response to cloudformation
+		report_err := h.reportStatus(req, false, "failed to decode payload for metricsconfigurer")
+		if report_err != nil {
+			return nil, fmt.Errorf("failed to report status to cloudformation: %w, for error %w", report_err, err)
+		}
+		return nil, fmt.Errorf("failed to decode payload for metricsconfigurer: %w", err)
+	}
+	return &req, nil
+}
+
+func (h Handler) parseResponse(bodyBytes []byte, req *Request) ([]types.MetricStreamFilter, error) {
+	logger := h.Logger
+	var result GraphQLResponse
+	err := json.Unmarshal(bodyBytes, &result)
+	if err != nil {
+		report_err := h.reportStatus(*req, false, fmt.Sprintf("error unmarshalling JSON from GQL response %s", err))
+		if report_err != nil {
+			return nil, fmt.Errorf("failed to report status to cloudformation: %w, for error %w", report_err, err)
+		}
+		return nil, fmt.Errorf("error unmarshalling JSON from GQL response: %w", err)
+	}
+
+	logger.V(4).Info("response from observe api", "result", result)
+
+	var metricSelection []MetricsListItem
+
+	variablesList := result.Data.Datasource.Variables
+
+	var targetVariable *Variable
+
+	// Iterate through the variables and find one with the name "Metrics"
+	// This is where we should find the metrics configuration
+	for _, variable := range variablesList {
+		if variable.Name == "Metrics" {
+			targetVariable = &variable
+			break
+		}
+	}
+
+	if targetVariable == nil {
+		report_err := h.reportStatus(*req, false, "metrics variable not set in datasource")
+		if report_err != nil {
+			return nil, fmt.Errorf("failed to report status to cloudformation: %w, for error %w", report_err, err)
+		}
+		return nil, fmt.Errorf("metrics variable not set in datasource, %+v", result)
+	}
+
+	metricSelection = targetVariable.Details.AwsCollectionStackConfig.AwsServiceMetricsList
+
+	MetricsFilters := convertToMetricStreamFilters(metricSelection)
+	return MetricsFilters, nil
+}
+
+func convertToMetricStreamFilters(MetricsList []MetricsListItem) []types.MetricStreamFilter {
+
+	// to ensure that an empty slice is returned instead of nil when `MetricsList` is empty
+	metricsFilters := make([]types.MetricStreamFilter, 0)
+	for _, service := range MetricsList {
+		metricsFilters = append(metricsFilters, types.MetricStreamFilter{
+			Namespace:   &service.Namespace,
+			MetricNames: service.MetricNames,
+		})
+	}
+
+	return metricsFilters
+}
+
+func (h Handler) getSecretValue(ctx context.Context, req *Request, cfg aws.Config) (*string, error) {
+
+	secretName := h.SecretName
+	svc := secretsmanager.NewFromConfig(cfg)
+	secretValue, err := svc.GetSecretValue(ctx, &secretsmanager.GetSecretValueInput{
+		SecretId: &secretName,
+	})
+	if err != nil {
+		report_err := h.reportStatus(*req, false, fmt.Sprintf("failed to retrieve secret value %s", err))
+		if report_err != nil {
```

Comment:
mentioned this before, seems like we can just wrap these 4 lines into a helper function, they're fairly repetitive and it can just take in the specific error message as a param